src/main/scala/com/databricks/spark/avro/AvroRelation.scala

```diff
@@ -31,24 +31,32 @@ import org.apache.avro.Schema.Type._
 
 import org.apache.hadoop.fs.{FileSystem, Path}
 
-import org.apache.spark.sql._
-import org.apache.spark.sql.sources.TableScan
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.sources.{BaseRelation, InsertableRelation, TableScan}
 
 
-case class AvroRelation(location: String, minPartitions: Int = 0)
-    (@transient val sqlContext: SQLContext) extends TableScan {
+case class AvroRelation(
+    location: String,
+    userSpecifiedSchema: Option[StructType],
+    minPartitions: Int = 0) (@transient val sqlContext: SQLContext)
+    extends BaseRelation with TableScan with InsertableRelation {
   var avroSchema: Schema = null
 
   override val schema = {
-    val fileReader = newReader()
-    avroSchema = fileReader.getSchema
-    val convertedSchema = SchemaConverters.toSqlType(fileReader.getSchema).dataType match {
-      case s: StructType => s
-      case other =>
-        sys.error(s"Avro files must contain Records to be read, type $other not supported")
+    if (userSpecifiedSchema != None) {
+      userSpecifiedSchema.get
+    } else {
+      val fileReader = newReader()
+      avroSchema = fileReader.getSchema
+      val convertedSchema = SchemaConverters.toSqlType(fileReader.getSchema).dataType match {
+        case s: StructType => s
+        case other =>
+          sys.error(s"Avro files must contain Records to be read, type $other not supported")
+      }
+      fileReader.close()
```

Comment:
no, to make sure we close the reader, otherwise we may leak the input stream if an exception is thrown